File: ui/lib/hygiene.ts
//ambient dev tool that watches what you do and updates your PM tickets automatically, boosting developer productivity
//
// Maps the triage reason codes stored in pm_task_curation.reasons_json into the
// human hint + the in-app fix the Tasks view renders. Mirrors the Rust engine's
// TriageReason::hint() and TriageReason::fix() — kept in sync by hand.

export interface HygieneFix {
  control: string // date_picker | assign_self | edit_text | edit_checklist | pick_parent | edit_labels | pick_priority | number_input
  field: string // tracker field the write-back targets
  label: string // button text
  ai: boolean // suggested value comes from AI (wired later)
}

export type Severity = 'must_fix' | 'optional'

export interface HygieneIssue {
  code: string
  hint: string
  fix: HygieneFix | null // null ⇒ handled at ticket level (close / snooze) or descriptive
  severity: Severity // must_fix: Meridian needs it (due date / description / title) — the rest is optional
}

// Must-fix = the fields Meridian needs to track a ticket accurately. Everything
// else is good hygiene the dev can address at leisure on the cleanup page.
const MUST_FIX = new Set([
  'missing_description',
  'thin_description',
  'vague_title',
  'missing_due_date',
])

function reasonSeverity(code: string): Severity {
  return MUST_FIX.has(code) ? 'must_fix' : 'optional'
}

export interface Hygiene {
  bucket: string // ready | needs_detail | looks_stale | not_sure
  issues: HygieneIssue[] // fixable defects (excludes purely-active reasons)
  decision: string | null // keep | excluded | snoozed
}

function reasonHint(code: string, d: Record<string, number> | undefined): string {
  switch (code) {
    case 'in_progress': return 'In progress on the board.'
    case 'due_soon': return (d?.in_days ?? 1) <= 0 ? 'Due today.' : `Due in ${d?.in_days} day(s).`
    case 'in_sprint': return 'In the active sprint.'
    case 'start_date_reached': return 'Its start date has passed.'
    case 'missing_description': return 'No description — nothing to match your work against.'
    case 'thin_description': return `Description is only ${d?.chars} characters.`
    case 'vague_title': return 'Title is generic — make it specific.'
    case 'no_context_anchor': return 'Not linked to an epic or parent.'
    case 'missing_due_date': return "No due date — add one so Meridian knows when it's live."
    case 'missing_assignee': return 'No assignee — who owns this?'
    case 'missing_labels': return 'No labels — add one to categorise it.'
    case 'missing_priority': return 'No priority set.'
    case 'missing_estimate': return 'No estimate — add story points.'
    case 'missing_acceptance_criteria': return "No acceptance criteria — define what 'done' means."
    case 'no_activity_since': return `No board activity in ${d?.days} days.`
    case 'not_started': return 'Still in a not-started column.'
    case 'no_due_date': return 'No due date set.'
    case 'overdue_long': return `Overdue by ${d?.by_days} days with no movement.`
    case 'far_future_due': return `Not due for ${d?.in_days} days — planned, not current work.`
    case 'not_in_sprint': return 'Not in any sprint.'
    case 'already_done': return 'Already marked done.'
    case 'no_activity_signal': return "Open, but nothing yet says it's active."
    case 'unreadable_updated_at': return "Couldn't read its last-updated time."
    default: return code
  }
}

function reasonFix(code: string): HygieneFix | null {
  switch (code) {
    case 'missing_description': return { control: 'edit_text', field: 'description', label: 'Add a description', ai: true }
    case 'thin_description': return { control: 'edit_text', field: 'description', label: 'Expand the description', ai: true }
    case 'vague_title': return { control: 'edit_text', field: 'summary', label: 'Make the title specific', ai: true }
    case 'no_context_anchor': return { control: 'pick_parent', field: 'parent', label: 'Link to an epic or parent', ai: false }
    case 'missing_due_date': return { control: 'date_picker', field: 'duedate', label: 'Add a due date', ai: false }
    case 'missing_assignee': return { control: 'assign_self', field: 'assignee', label: 'Assign to me', ai: false }
    case 'missing_labels': return { control: 'edit_labels', field: 'labels', label: 'Add a label', ai: false }
    case 'missing_priority': return { control: 'pick_priority', field: 'priority', label: 'Set priority', ai: false }
    case 'missing_estimate': return { control: 'number_input', field: 'story_points', label: 'Add an estimate', ai: false }
    case 'missing_acceptance_criteria': return { control: 'edit_checklist', field: 'acceptance_criteria', label: 'Add acceptance criteria', ai: true }
    default: return null // stale / active / descriptive reasons aren't per-field fixes
  }
}

interface RawReason { code: string; detail?: Record<string, number> }

/** Parse a reasons_json blob into the fixable hygiene issues (drops active/descriptive). */
export function parseIssues(reasonsJson: string | null): HygieneIssue[] {
  if (!reasonsJson) return []
  let raw: RawReason[]
  try { raw = JSON.parse(reasonsJson) } catch { return [] }
  return raw
    .map(r => ({
      code: r.code,
      hint: reasonHint(r.code, r.detail),
      fix: reasonFix(r.code),
      severity: reasonSeverity(r.code),
    }))
    .filter(i => i.fix !== null)
}

/** True if any issue is must-fix — drives the Tasks-page banner. */
export function hasMustFix(issues: HygieneIssue[]): boolean {
  return issues.some(i => i.severity === 'must_fix')
}

File: meridian-core/src/util/hygiene.rs
//ambient dev tool that watches what you do and updates your PM tickets automatically, boosting developer productivity
//! Board-hygiene mapping — a faithful port of `ui/lib/hygiene.ts` (`parseIssues`
//! + reason hint/fix/severity). Maps the triage reason codes the daemon stores
//! in `pm_task_curation.reasons_json` into the human hint + the in-app fix the
//! Tasks view renders. Mirrors the Rust engine's `TriageReason::{hint,fix}` —
//! kept in sync by hand (same as the TS did).

use serde::Serialize;
use serde_json::{Map, Value};

#[derive(Debug, Clone, Serialize)]
pub struct HygieneFix {
    pub control: String,
    pub field: String,
    pub label: String,
    pub ai: bool,
}

#[derive(Debug, Clone, Serialize)]
pub struct HygieneIssue {
    pub code: String,
    pub hint: String,
    pub fix: Option<HygieneFix>,
    /// "must_fix" | "optional"
    pub severity: String,
}

#[derive(Debug, Clone, Serialize)]
pub struct Hygiene {
    pub bucket: String,
    pub issues: Vec<HygieneIssue>,
    pub decision: Option<String>,
}

/// Fields Meridian needs to track a ticket accurately (must-fix); the rest is
/// optional good hygiene.
const MUST_FIX: [&str; 5] = [
    "missing_description",
    "thin_description",
    "vague_title",
    "missing_due_date",
    "overdue",
];

fn reason_severity(code: &str) -> &'static str {
    if MUST_FIX.contains(&code) {
        "must_fix"
    } else {
        "optional"
    }
}

/// A detail number formatted as JS would interpolate it (`${d?.key}`): the
/// integer/float value, or the literal "undefined" when the key is absent.
fn jsnum(detail: Option<&Map<String, Value>>, key: &str) -> String {
    match detail.and_then(|d| d.get(key)) {
        Some(v) if v.is_i64() => v.as_i64().unwrap().to_string(),
        Some(v) if v.is_u64() => v.as_u64().unwrap().to_string(),
        Some(v) if v.is_f64() => v.as_f64().unwrap().to_string(),
        _ => "undefined".to_string(),
    }
}

fn reason_hint(code: &str, detail: Option<&Map<String, Value>>) -> String {
    match code {
        "in_progress" => "In progress on the board.".to_string(),
        "due_soon" => {
            let in_days = detail
                .and_then(|d| d.get("in_days"))
                .and_then(|v| v.as_i64())
                .unwrap_or(1);
            if in_days <= 0 {
                "Due today.".to_string()
            } else {
                format!("Due in {} day(s).", jsnum(detail, "in_days"))
            }
        }
        "in_sprint" => "In the active sprint.".to_string(),
        "start_date_reached" => "Its start date has passed.".to_string(),
        "missing_description" => "No description — nothing to match your work against.".to_string(),
        "thin_description" => format!("Description is only {} characters.", jsnum(detail, "chars")),
        "vague_title" => "Title is generic — make it specific.".to_string(),
        "no_context_anchor" => "Not linked to an epic or parent.".to_string(),
        "missing_due_date" => "No due date — add one so Meridian knows when it's live.".to_string(),
        "overdue" => {
            let by = detail
                .and_then(|d| d.get("by_days"))
                .and_then(|v| v.as_i64())
                .unwrap_or(0);
            format!("Overdue by {by} day(s) — reschedule or close it.")
        }
        "missing_assignee" => "No assignee — who owns this?".to_string(),
        "missing_labels" => "No labels — add one to categorise it.".to_string(),
        "missing_priority" => "No priority set.".to_string(),
        "missing_estimate" => "No estimate — add story points.".to_string(),
        "missing_acceptance_criteria" => {
            "No acceptance criteria — define what 'done' means.".to_string()
        }
        "no_activity_since" => format!("No board activity in {} days.", jsnum(detail, "days")),
        "not_started" => "Still in a not-started column.".to_string(),
        "no_due_date" => "No due date set.".to_string(),
        "overdue_long" => format!(
            "Overdue by {} days with no movement.",
            jsnum(detail, "by_days")
        ),
        "far_future_due" => format!(
            "Not due for {} days — planned, not current work.",
            jsnum(detail, "in_days")
        ),
        "not_in_sprint" => "Not in any sprint.".to_string(),
        "already_done" => "Already marked done.".to_string(),
        "no_activity_signal" => "Open, but nothing yet says it's active.".to_string(),
        "unreadable_updated_at" => "Couldn't read its last-updated time.".to_string(),
        other => other.to_string(),
    }
}

fn fix(control: &str, field: &str, label: &str, ai: bool) -> HygieneFix {
    HygieneFix {
        control: control.to_string(),
        field: field.to_string(),
        label: label.to_string(),
        ai,
    }
}

fn reason_fix(code: &str) -> Option<HygieneFix> {
    match code {
        "missing_description" => Some(fix("edit_text", "description", "Add a description", true)),
        "thin_description" => Some(fix(
            "edit_text",
            "description",
            "Expand the description",
            true,
        )),
        "vague_title" => Some(fix("edit_text", "summary", "Make the title specific", true)),
        "no_context_anchor" => Some(fix(
            "pick_parent",
            "parent",
            "Link to an epic or parent",
            false,
        )),
        "missing_due_date" => Some(fix("date_picker", "duedate", "Add a due date", false)),
        "overdue" => Some(fix("date_picker", "duedate", "Reschedule due date", false)),
        "missing_assignee" => Some(fix("assign_self", "assignee", "Assign to me", false)),
        "missing_labels" => Some(fix("edit_labels", "labels", "Add a label", false)),
        "missing_priority" => Some(fix("pick_priority", "priority", "Set priority", false)),
        "missing_estimate" => Some(fix(
            "number_input",
            "story_points",
            "Add an estimate",
            false,
        )),
        "missing_acceptance_criteria" => Some(fix(
            "edit_checklist",
            "acceptance_criteria",
            "Add acceptance criteria",
            true,
        )),
        _ => None,
    }
}

/// Parse a reasons_json blob into the fixable hygiene issues (drops active/
/// descriptive reasons with no fix, and any the dev chose to ignore).
pub fn parse_issues(reasons_json: Option<&str>, ignored_json: Option<&str>) -> Vec<HygieneIssue> {
    let Some(reasons_json) = reasons_json else {
        return Vec::new();
    };
    let Ok(raw) = serde_json::from_str::<Vec<Value>>(reasons_json) else {
        return Vec::new();
    };
    let ignored: Vec<String> = ignored_json
        .and_then(|s| serde_json::from_str::<Vec<String>>(s).ok())
        .unwrap_or_default();

    raw.into_iter()
        .filter_map(|r| {
            let code = r.get("code")?.as_str()?.to_string();
            let detail = r.get("detail").and_then(|d| d.as_object());
            let fix = reason_fix(&code);
            // Mirror the TS filter: keep only fixable, non-ignored reasons.
            if fix.is_none() || ignored.contains(&code) {
                return None;
            }
            Some(HygieneIssue {
                hint: reason_hint(&code, detail),
                severity: reason_severity(&code).to_string(),
                fix,
                code,
            })
        })
        .collect()
}

#[cfg(test)]
mod tests {
    use super::*;

    #[test]
    fn keeps_only_fixable_reasons() {
        // in_progress / due_soon have no fix → dropped; missing_due_date stays.
        let reasons = r#"[
            {"code":"in_progress"},
            {"code":"due_soon","detail":{"in_days":2}},
            {"code":"missing_due_date"}
        ]"#;
        let issues = parse_issues(Some(reasons), None);
        assert_eq!(issues.len(), 1);
        assert_eq!(issues[0].code, "missing_due_date");
        assert!(issues[0].fix.is_some());
    }

    #[test]
    fn ignored_codes_are_filtered_out() {
        let reasons = r#"[{"code":"missing_labels"},{"code":"missing_priority"}]"#;
        let issues = parse_issues(Some(reasons), Some(r#"["missing_labels"]"#));
        assert_eq!(issues.len(), 1);
        assert_eq!(issues[0].code, "missing_priority");
    }

    #[test]
    fn severity_splits_must_fix_from_optional() {
        let reasons = r#"[{"code":"missing_due_date"},{"code":"missing_labels"}]"#;
        let issues = parse_issues(Some(reasons), None);
        let by = |c: &str| issues.iter().find(|i| i.code == c).unwrap();
        assert_eq!(by("missing_due_date").severity, "must_fix");
        assert_eq!(by("missing_labels").severity, "optional");
    }

    #[test]
    fn thin_description_hint_carries_the_char_count() {
        let issues = parse_issues(
            Some(r#"[{"code":"thin_description","detail":{"chars":12}}]"#),
            None,
        );
        assert_eq!(issues[0].hint, "Description is only 12 characters.");
    }

    #[test]
    fn fix_shape_matches_the_control_mapping() {
        let issues = parse_issues(Some(r#"[{"code":"missing_assignee"}]"#), None);
        let fix = issues[0].fix.as_ref().unwrap();
        assert_eq!(fix.control, "assign_self");
        assert_eq!(fix.field, "assignee");
        assert!(!fix.ai);
    }

    #[test]
    fn empty_or_malformed_yields_no_issues() {
        assert!(parse_issues(None, None).is_empty());
        assert!(parse_issues(Some(""), None).is_empty());
        assert!(parse_issues(Some("not json"), None).is_empty());
        assert!(parse_issues(Some("[]"), None).is_empty());
    }
}
